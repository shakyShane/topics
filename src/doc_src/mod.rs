pub mod doc_src;
pub use doc_src::*;
