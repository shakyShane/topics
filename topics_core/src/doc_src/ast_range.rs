pub trait AstRangeImpl {
    fn ast_range(&self) -> AstRange;
}

///
/// An AST (tree-like) range selector
///
/// ```rust
/// # fn main() -> anyhow::Result<()> {
/// # use comrak::Arena;
/// # use topics_core::doc_src::MdElements;
/// let input = "# heading";
/// let arena = Arena::new();
/// let md_elements = MdElements::new(input, &arena);
///
/// // tuple of (vec![0], 1) is path + len
/// let html = md_elements.as_html((vec![0], 1));
/// assert_eq!(html, String::from("<h1>heading</h1>\n"));
/// # Ok(())
/// # }
/// ```
#[derive(Debug, Clone, Default)]
pub struct AstRange {
    pub ast_path: Vec<usize>,
    pub ast_len: usize,
}

impl AstRange {
    pub fn range(path: &[usize], len: usize) -> Self {
        Self {
            ast_path: path.to_vec(),
            ast_len: len,
        }
    }
}

impl AstRangeImpl for AstRange {
    fn ast_range(&self) -> AstRange {
        self.clone()
    }
}

impl AstRangeImpl for (Vec<usize>, usize) {
    fn ast_range(&self) -> AstRange {
        AstRange {
            ast_path: self.0.to_vec(),
            ast_len: self.1,
        }
    }
}

impl AstRangeImpl for (&'_ Vec<usize>, usize) {
    fn ast_range(&self) -> AstRange {
        AstRange {
            ast_path: self.0.to_vec(),
            ast_len: self.1,
        }
    }
}
