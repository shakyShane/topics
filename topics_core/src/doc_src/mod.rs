pub mod doc_src;
pub mod yaml_doc_src;

pub use doc_src::*;
pub use yaml_doc_src::*;
