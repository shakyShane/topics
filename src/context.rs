use crate::doc::{Doc, DocResult};
use crate::opt::Opt;
use crate::print::Print;
use std::path::PathBuf;

#[derive(Debug, Default)]
pub struct Context {
    pub opts: Opt,
}

impl Context {
    pub fn from_opts(opt: &Opt) -> Self {
        Self {
            opts: (*opt).clone(),
        }
    }
    pub fn from_vec(input: &[&str]) -> Self {
        Self::from_opts(&Opt::from_vec(input))
    }
    pub fn join_path(&self, pb: impl Into<PathBuf>) -> PathBuf {
        self.opts.cwd.join_path(pb)
    }
    pub fn read_docs_split(&self) -> (Vec<DocResult<Doc>>, Vec<DocResult<Doc>>) {
        self.read_docs().into_iter().partition(|a| match a {
            Ok(doc) => doc.errors.is_empty(),
            Err(_) => false,
        })
    }
    pub fn read_docs(&self) -> Vec<DocResult<Doc>> {
        self.opts
            .files
            .iter()
            .map(|pb| Doc::from_path_buf(pb, &self))
            .collect()
    }
    pub fn cwd(&self) -> PathBuf {
        self.opts.cwd.0.clone()
    }
}

impl Print for Context {
    fn print(&self, doc: &Doc, ctx: &Context) -> anyhow::Result<()> {
        self.opts.print_kind.print(&doc, &ctx)
    }

    fn print_welcome(&self, docs: &Vec<DocResult<Doc>>, ctx: &Context) -> anyhow::Result<()> {
        self.opts.print_kind.print_welcome(&docs, &ctx)
    }

    fn print_all(&self, docs: &Vec<DocResult<Doc>>, ctx: &Context) -> anyhow::Result<()> {
        self.opts.print_kind.print_all(&docs, &ctx)
    }
    fn print_errors(&self, docs: &Vec<DocResult<Doc>>, ctx: &Context) -> anyhow::Result<()> {
        self.opts.print_kind.print_errors(&docs, &ctx)
    }
}
