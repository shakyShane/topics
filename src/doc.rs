use crate::command::Command;
use crate::context::Context;
use crate::dependency::DependencyCheck;
use crate::doc_src::DocSource;
use crate::instruction::Instruction;
use crate::item::Item;
use crate::topic::Topic;

use std::collections::HashMap;
use std::path::PathBuf;
use std::str::FromStr;

#[derive(Debug, Default)]
pub struct Doc {
    pub input_file: PathBuf,
    pub sources: Vec<DocSource>,
    pub topics: HashMap<String, Topic>,
    pub instructions: HashMap<String, Instruction>,
    pub dep_checks: HashMap<String, DependencyCheck>,
    pub commands: HashMap<String, Command>,
}

impl Doc {
    pub fn from_path_buf(pb: &PathBuf, ctx: &Context) -> anyhow::Result<Self> {
        let doc_src = DocSource::from_path_buf(&pb, ctx)?;
        Self::from_doc_src(&pb, doc_src, &ctx)
    }
    pub fn from_doc_src(
        pb: &PathBuf,
        doc_srcs: Vec<DocSource>,
        _ctx: &Context,
    ) -> anyhow::Result<Self> {
        let mut doc = Doc::default();
        doc.input_file = pb.clone();
        doc.sources = doc_srcs;
        for src in &doc.sources {
            let item: Item = serde_yaml::from_str(&src.content)?;
            match item {
                Item::Command(cmd) => {
                    doc.commands.insert(cmd.name.clone(), cmd.clone());
                }
                Item::FileExistsCheck(_) => {}
                Item::DependencyCheck(dc) => {
                    doc.dep_checks.insert(dc.name.clone(), dc.clone());
                }
                Item::Instruction(inst) => {
                    doc.instructions.insert(inst.name.clone(), inst.clone());
                }
                Item::HostEntriesCheck(_) => {}
                Item::Topic(t) => {
                    doc.topics.insert(t.name.clone(), t.clone());
                }
            };
        }
        Ok(doc)
    }
}

impl FromStr for Doc {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let doc_srcs = DocSource::parse(s)?;
        Doc::from_doc_src(&PathBuf::new(), doc_srcs, &Default::default())
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_single_doc() -> anyhow::Result<()> {
        let input = r#"
kind: Topic
name: Run screen shot tests
deps:
  - global-node
  - global-yarn
steps:
  - github-checkin
"#;
        let doc = Doc::from_str(input)?;
        assert_eq!(doc.sources.len(), 1);
        Ok(())
    }

    #[test]
    fn test_multi_doc() -> anyhow::Result<()> {
        let input = r#"
---
kind: Topic
name: Run screen shot tests
deps:
  - global-node
  - global-yarn
steps:
  - github-checkin
---
---
---
kind: Instruction
instruction: help me!
name: help-me-instruction
---
---
"#;
        let doc = Doc::from_str(input)?;
        assert_eq!(doc.sources.len(), 2);
        Ok(())
    }
}
