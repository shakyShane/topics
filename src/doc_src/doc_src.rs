use crate::context::Context;

use std::path::PathBuf;

#[derive(Debug, Default)]
pub struct DocSource {
    pub line_start: usize,
    pub line_end: usize,
    pub content: String,
}

impl DocSource {
    pub fn from_path_buf(pb: &PathBuf, ctx: &Context) -> anyhow::Result<Vec<Self>> {
        let abs = ctx.join_path(pb);
        let file_str = std::fs::read_to_string(&abs).map_err(|e| DocSrcError::FileRead {
            pb: pb.clone(),
            abs: abs.clone(),
            original: e,
        })?;
        Self::parse(&file_str)
    }
    pub fn parse(str: &str) -> anyhow::Result<Vec<Self>> {
        let mut output: Vec<Self> = vec![];
        let split = str.lines().collect::<Vec<&str>>();
        let mut peek = split.iter().enumerate().peekable();
        let mut start = 0;
        let mut end = 0;
        let mut end_line = 0;
        let mut docs: Vec<(usize, usize)> = vec![];
        while let Some((line, content)) = peek.next() {
            if content.starts_with("---") {
                let finish = line - 1;
                if start < finish {
                    docs.push((start, line));
                    end = line;
                }
                start = line + 1;
            } else if peek.peek().is_none() {
                end_line = line + 1;
                if end < start {
                    docs.push((start, line + 1))
                } else {
                    if end > start {
                        docs.push((start, end))
                    }
                }
            }
        }
        if start == 0 && end == 0 {
            docs.push((start, end_line))
        }

        for (start, end) in docs {
            let content = split[start..end].join("\n");
            output.push(Self {
                line_start: start,
                line_end: end,
                content,
            });
        }
        Ok(output)
    }
}

#[derive(Debug, thiserror::Error)]
enum DocSrcError {
    #[error(
    "FileRead error: could not read file `{}`\nFull path: {}",
    pb.display(),
    abs.display()
    )]
    FileRead {
        pb: PathBuf,
        abs: PathBuf,
        original: std::io::Error,
    },
}

#[cfg(test)]
mod test {

    use crate::context::Context;
    use crate::doc::Doc;
    use crate::doc_src::DocSource;
    use std::env::current_dir;

    #[test]
    fn test_fixture_file() -> anyhow::Result<()> {
        let ctx = Context::from_vec(&[]);
        let pb = current_dir()?.join("fixtures2/topics.yaml");
        let d = DocSource::from_path_buf(&pb, &ctx)?;
        insta::assert_debug_snapshot!(d);
        Ok(())
    }

    #[test]
    fn test_single_doc() -> anyhow::Result<()> {
        let input = r#"
kind: Topic
name: Run screen shot tests
deps:
  - global-node
  - global-yarn
steps:
  - github-checkin
"#;
        let doc = DocSource::parse(input)?;
        assert_eq!(doc.len(), 1);
        Ok(())
    }

    #[test]
    fn test_single_doc_with_prefix_padding() -> anyhow::Result<()> {
        let input = r#"
---
---
kind: Topic
name: Run screen shot tests
deps:
  - global-node
  - global-yarn
steps:
  - github-checkin
"#;
        let srcs = DocSource::parse(input)?;
        insta::assert_debug_snapshot!(srcs);
        Ok(())
    }

    #[test]
    fn test_single_doc_with_suffix_padding() -> anyhow::Result<()> {
        let input = r#"
kind: Topic
name: Run screen shot tests
deps:
  - global-node
  - global-yarn
steps:
  - github-checkin
  - |
    some other long string
---
---
"#;
        let srcs = DocSource::parse(input)?;
        insta::assert_debug_snapshot!(srcs);
        Ok(())
    }
}
