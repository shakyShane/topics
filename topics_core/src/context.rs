use crate::doc::{Doc, DocResult};
use crate::opt::Opt;
use std::path::PathBuf;

#[derive(Debug, Default)]
pub struct Context {
    pub opts: Opt,
}

impl Context {
    pub fn from_opts(opt: &Opt) -> Self {
        Self {
            opts: (*opt).clone(),
        }
    }
    #[cfg(test)]
    pub fn from_vec(input: &[&str]) -> Self {
        Self::from_opts(&Opt::from_vec(input))
    }
    pub fn join_path(&self, pb: impl Into<PathBuf>) -> PathBuf {
        self.opts.cwd.join_path(pb)
    }
    pub fn read_docs_split(&self, files: &[PathBuf]) -> (Vec<DocResult<Doc>>, Vec<DocResult<Doc>>) {
        self.read_docs(&files).into_iter().partition(|a| match a {
            Ok(doc) => doc.errors.is_empty(),
            Err(_) => false,
        })
    }
    pub fn read_docs(&self, files: &[PathBuf]) -> Vec<DocResult<Doc>> {
        files
            .iter()
            .map(|pb| Doc::from_path_buf(pb, &self))
            .collect()
    }
    #[cfg(test)]
    pub fn read_docs_unwrapped(&self, files: &[PathBuf]) -> Vec<Doc> {
        files
            .iter()
            .map(|pb| Doc::from_path_buf(pb, &self).expect("read_docs_unwrapped"))
            .collect()
    }
    pub fn _cwd(&self) -> PathBuf {
        self.opts.cwd.0.clone()
    }
}
