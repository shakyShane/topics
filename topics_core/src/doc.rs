use crate::doc_err::DocError;
use crate::doc_src::{from_serde_yaml_error, parse_md_str, DocSource, TomlError};
use crate::items::item::Item;
use crate::items::ItemWrap;
use crate::{context::Context, items::Topic};
use std::path::PathBuf;

#[derive(Debug, Default)]
pub struct Doc {
    pub source: DocSource,
    pub items: Vec<ItemTracked>,
    pub errors: Vec<DocError>,
}

#[derive(Debug, Clone)]
pub struct ItemTracked {
    pub item: Item,
    pub src_doc: DocSource,
    pub input_file: Option<PathBuf>,
}

pub type DocResult<T, E = DocError> = core::result::Result<T, E>;

impl Doc {
    pub fn topics(&self) -> Vec<Topic> {
        self.items
            .iter()
            .filter_map(|item| match item {
                ItemTracked {
                    item: Item::Topic(topic),
                    ..
                } => Some(topic.clone()),
                _ => None,
            })
            .collect()
    }
    pub fn topic_names(&self) -> Vec<&str> {
        self.items
            .iter()
            .filter_map(|item| match item {
                ItemTracked {
                    item: Item::Topic(topic),
                    ..
                } => Some(topic.name.as_str()),
                _ => None,
            })
            .collect()
    }
    pub fn topic_by_name(&self, name: &str) -> Option<&Topic> {
        self.items.iter().find_map(|item| match item {
            ItemTracked {
                item: Item::Topic(topic),
                ..
            } => {
                if topic.name == name {
                    Some(topic)
                } else {
                    None
                }
            }
            _ => None,
        })
    }
    pub fn from_path_buf(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        let doc_src = match pb.extension() {
            None => todo!("what to handle here?"),
            Some(os_str) => match os_str.to_str() {
                None => todo!("what to handle here?"),
                Some("yaml") | Some("yml") => DocSource::yaml(&pb, ctx)?,
                Some("toml") => DocSource::toml(&pb, ctx)?,
                Some("md") | Some("markdown") => DocSource::md(&pb, ctx)?,
                Some(_other) => return Err(DocError::NotSupported(pb.clone())),
            },
        };
        Self::from_doc_src(&pb, doc_src, &ctx)
    }
    pub fn from_doc_src(_pb: &PathBuf, doc_src: DocSource, _ctx: &Context) -> DocResult<Self> {
        let mut doc = Doc {
            source: doc_src,
            ..Default::default()
        };
        match &doc.source {
            DocSource::Yaml(yaml_doc) => {
                for src in &yaml_doc.doc_src_items.items {
                    let item: Result<Item, DocError> = serde_yaml::from_str(&src.content)
                        .map_err(|err| from_serde_yaml_error(&doc, &src, &err));
                    match item {
                        Err(doc_err) => {
                            doc.errors.push(doc_err);
                        }
                        Ok(item) => {
                            doc.items.push(ItemTracked {
                                item,
                                src_doc: DocSource::Yaml((*yaml_doc).clone()),
                                input_file: yaml_doc.input_file.clone(),
                            });
                        }
                    };
                }
            }
            DocSource::Toml(toml_doc) => {
                let items = one_or_many_toml(&toml_doc.file_content).map_err(|err| TomlError {
                    doc: &doc,
                    toml_err: err,
                })?;
                for item in items {
                    doc.items.push(ItemTracked {
                        item,
                        src_doc: DocSource::Toml((*toml_doc).clone()),
                        input_file: toml_doc.input_file.clone(),
                    });
                }
            }
            DocSource::Md(md_doc) => {
                for src in &md_doc.doc_src_items.items {
                    let items = parse_md_str(&src.content)?;
                    for item in items {
                        if let Item::Topic(topic) = &item {
                            for item_wrap in topic.deps.iter().chain(topic.steps.iter()) {
                                if let ItemWrap::Item(item) = item_wrap {
                                    doc.items.push(ItemTracked {
                                        item: item.clone(),
                                        src_doc: DocSource::Md((*md_doc).clone()),
                                        input_file: md_doc.input_file.clone(),
                                    });
                                }
                            }
                        }
                        doc.items.push(ItemTracked {
                            item,
                            src_doc: DocSource::Md((*md_doc).clone()),
                            input_file: md_doc.input_file.clone(),
                        });
                    }
                }
            }
        }
        Ok(doc)
    }
}

fn one_or_many_toml(input: &str) -> Result<Vec<Item>, toml::de::Error> {
    #[derive(Debug, serde::Deserialize)]
    struct TempItems {
        item: Vec<Item>,
    }
    toml::from_str::<TempItems>(input)
        .or_else(|err| {
            if err
                .to_string()
                .contains("missing field `item` at line 1 column 1")
            {
                toml::from_str::<Item>(input).map(|item| TempItems { item: vec![item] })
            } else {
                Err(err)
            }
        })
        .map(|temp| temp.item)
}

#[test]
fn test_toml() -> anyhow::Result<()> {
    let input = r#"
[[item]]
kind = "Topic"
name = "Run unit tests"
steps = [
    "Install root-level dependencies",
    "Run unit tests command"
]
deps = [
    "install nodejs"
]

[[item]]
kind = "Command"
name = "Run unit tests command"
cwd = "."
command = """
echo hello world!
"""

[[item]]
kind = "Command"
name = "Install root-level dependencies"
cwd = "."
command = """
echo hello world!
"""

[[item]]
kind = "DependencyCheck"
name = "install nodejs"
verify = """
node -v
"""
    "#;
    #[derive(Debug, serde::Deserialize)]
    struct Items {
        item: Vec<Item>,
    }
    let items: Items = toml::from_str(input)?;
    assert_eq!(items.item.len(), 4);
    Ok(())
}

#[test]
fn test_toml_single() {
    let input = r#"
kind = "Command"
cwd = "."
name = "Run unit tests command"
command = """
echo "About to install ${MIN_VERSION}"
"""

env.HELLO = { from = "env vars", key = "minimum_skaffold_version" }

    "#;
    let items = one_or_many_toml(input);
    println!("{:?}", items);
}
