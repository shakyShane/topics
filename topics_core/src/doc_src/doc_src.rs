use crate::context::Context;
use crate::doc::DocResult;
use crate::doc_src::{MdDocSource, TomlDocSource, YamlDocSource};

use multi_doc::SingleDoc;
use std::path::PathBuf;

pub trait DocSrcImpl: Sized {
    fn from_path_buf(pb: &PathBuf, ctx: &Context) -> DocResult<Self>;
}

#[derive(Debug, Clone)]
pub enum DocSource {
    Yaml(YamlDocSource),
    Toml(TomlDocSource),
    Md(MdDocSource),
}

impl DocSource {
    pub fn file(&self) -> Option<PathBuf> {
        match self {
            DocSource::Yaml(yaml_doc) => yaml_doc.input_file.clone(),
            DocSource::Toml(toml_doc) => toml_doc.input_file.clone(),
            DocSource::Md(md_doc) => md_doc.input_file.clone(),
        }
    }
    pub fn content(&self) -> &str {
        match self {
            DocSource::Yaml(yaml_doc) => yaml_doc.file_content.as_str(),
            DocSource::Toml(toml_doc) => toml_doc.file_content.as_str(),
            DocSource::Md(md_doc) => md_doc.file_content.as_str(),
        }
    }
    pub fn yaml(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        Ok(DocSource::Yaml(YamlDocSource::from_path_buf(&pb, ctx)?))
    }
    pub fn toml(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        Ok(DocSource::Toml(TomlDocSource::from_path_buf(&pb, ctx)?))
    }
    pub fn md(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        Ok(DocSource::Md(MdDocSource::from_path_buf(&pb, ctx)?))
    }
}

impl Default for DocSource {
    fn default() -> Self {
        Self::Yaml(YamlDocSource::default())
    }
}
