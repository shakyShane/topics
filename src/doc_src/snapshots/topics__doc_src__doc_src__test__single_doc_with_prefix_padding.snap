---
source: src/doc_src/doc_src.rs
expression: srcs

---
[
    DocSource {
        line_start: 3,
        line_end: 10,
        content: "kind: Topic\nname: Run screen shot tests\ndeps:\n  - global-node\n  - global-yarn\nsteps:\n  - github-checkin",
    },
]
